include/pybind11/cast.h

```diff
@@ -1006,6 +1006,27 @@ template <typename CharT> using is_std_char_type = any_of<
     std::is_same<CharT, wchar_t> /* std::wstring */
 >;
 
+
+template<typename T, typename U, enable_if_t<!std::is_integral<T>::value, int> = 0>
+bool invalid_convert_integral(U) {
+    return false;
+}
+
+template<typename T, typename U, enable_if_t<std::is_integral<T>::value && !std::is_unsigned<T>::value, int> = 0>
+bool invalid_convert_integral(U value) {
+    return sizeof(U) != sizeof(T)
+           && (   value < (U) (std::numeric_limits<T>::min)()
+               || value > (U) (std::numeric_limits<T>::max)()
+           );
+}
+
+template<typename T, typename U, enable_if_t<std::is_integral<T>::value && std::is_unsigned<T>::value, int> = 0>
+bool invalid_convert_integral(U value) {
+    return sizeof(U) != sizeof(T)
+           &&  value > (U) (std::numeric_limits<T>::max)();
```

Comment:
Is this completely correct?

`U == int8_t`, `T == uint32_t`, `value == 127` => `(U)numeric_limits::max() == (U)65535 == (U)0xff'ff'ff'ff == -1` => `127 > -1` and we end up saying this is invalid, when 127 can be represented in `uint32_t` losslessly.

Reply:
I noticed this too, by the time I finished with this I was pretty sure the (existing) logic is flawed. But I left it exactly the same for now, I just replicated what was there.

Reply:
The reason it worked is U is always large (python's long). But I've come up with a much simpler check that is inline.